src/module.rs

```diff
@@ -0,0 +1,110 @@
+use parking_lot::RwLock;
+use std::sync::Arc;
+
+use crate::field::{Field, FieldDef};
+use crate::ty::BuiltinType;
+use crate::utils::GrowableVec;
+use crate::vm::{FctId, TraitId, ImplId, TypeParam, VM, FileId};
+
+use dora_parser::lexer::position::Position;
+use dora_parser::interner::Name;
+use crate::vtable::VTableBox;
+use crate::class::ClassSize;
+
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub struct ModuleId(usize);
+
+impl ModuleId {
+    pub fn max() -> ModuleId {
+        ModuleId(usize::max_value())
+    }
+}
+
+impl From<ModuleId> for usize {
+    fn from(data: ModuleId) -> usize {
+        data.0
+    }
+}
+
+impl From<usize> for ModuleId {
+    fn from(data: usize) -> ModuleId {
+        ModuleId(data)
+    }
+}
+
+impl GrowableVec<RwLock<Module>> {
+    pub fn idx(&self, index: ModuleId) -> Arc<RwLock<Module>> {
+        self.idx_usize(index.0)
+    }
+}
+
+pub static DISPLAY_SIZE: usize = 6;
+
+#[derive(Debug)]
+pub struct Module {
+    pub id: ModuleId,
+    pub file: FileId,
+    pub pos: Position,
+    pub name: Name,
+    pub ty: BuiltinType,
+    pub parent_class: Option<ModuleId>,
+    pub internal: bool,
+    pub internal_resolved: bool,
+    pub has_constructor: bool,
+
+    pub constructor: Option<FctId>,
+    pub fields: Vec<Field>,
+    pub methods: Vec<FctId>,
+    pub virtual_fcts: Vec<FctId>,
+
+    pub traits: Vec<TraitId>,
+    pub impls: Vec<ImplId>,
+
+    pub type_params: Vec<TypeParam>,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct ModuleDefId(usize);
+
+impl ModuleDefId {
+    pub fn to_usize(self) -> usize {
+        self.0
+    }
+}
+
+impl From<usize> for ModuleDefId {
+    fn from(data: usize) -> ModuleDefId {
+        ModuleDefId(data)
+    }
+}
+
+impl GrowableVec<RwLock<ModuleDef>> {
+    pub fn idx(&self, index: ModuleDefId) -> Arc<RwLock<ModuleDef>> {
+        self.idx_usize(index.0)
+    }
+}
+
+#[derive(Debug)]
+pub struct ModuleDef {
+    pub id: ModuleDefId,
+    pub mod_id: Option<ModuleId>,
+    pub parent_id: Option<ModuleDefId>,
+    pub fields: Vec<FieldDef>,
+    pub size: ClassSize,
```

Comment:
As modules are supposed to have a singleton instance, I think it would make sense to store its size here (but it could be that I'm misunderstanding the field).

Reply:
Ok, thanks for the explanation. I was under the impression that we could avoid the actual singleton instance in the runtime (AFAIU we might still be able to do this when a module doesn't have a parent class). But since we can pass module instances to functions, this doesn't seem to be the case in general.

We might not need `ModuleDef` though, since a module isn't generic. `ClassDef` is a really bad name, it is necessary since classes can be generic and has different sizes/field types depending on the type parameters.

The GC/runtime use the vtable/classdef pointer to determine the type of objects and determine its size. Everything on the heap needs a ClassDef. Not sure if we should just create a `ClassDef` for the module and rename `ClassDef` to `Shape` or something, that it also matches the module instance.